### ISSUE TITLE
Getting email from JWT issued by Azure AD

### ISSUE THREAD

**Issue by summerwind** [coreos/dex#504]
Is there a way to change the claim to get the email in OIDC Connector?

I'd like to use dex with Azure AD. But JWT issued by Azure AD does not contain a "email" claim. Email is in "upn" or "unique_name" claims. So currently dex does not get the email fron JWT issued by Azure AD.

The information of claims in Azure AD is here:
https://azure.microsoft.com/en-us/documentation/articles/guidance-multitenant-identity-claims/

**Comment by sym3tri** [comment]
The root issue here is that Azure has decided to use non-standard claims in their JWTs[1]. They should use `email` b/c that is what's in the spec. Claims named "upn" and "unique_name" are not in the spec. 

This opens up a bigger philosophical discussion over weather or not we support non-standard OIDC providers, and if so how we do overrides. I can think of two possibilities:
  1. A map of upstream provider cliaim ID to Dex claim id
  2. We stuff a map of all upstream claims into a new claim (this may have other implications)

Also this is handled by go-oidc and should be moved here: https://github.com/coreos/go-oidc

[1] http://openid.net/specs/openid-connect-core-1_0.html#StandardClaims

**Comment by summerwind** [comment]
Thank you for your comment. As you said, Azure's JWT contains the non-standard claims. This is the root issue.

According to the documentation, there is a way to normalize claims if we use multiple IdP as dex.
https://azure.microsoft.com/en-us/documentation/articles/guidance-multitenant-identity-claims/#claims-transformations

> Claims normalization, or making claims consistent across users. This is particularly relevant if you are getting claims from multiple IDPs, which might use different claim types for similar information. For example, Azure AD sends a "upn" claim that contains the user's email. Other IDPs might send an "email" claim. The following code converts the "upn" claim into an "email" claim:

Adding normalization feature to go-oidc will make more complicated. I thought that  there is a way to adding the settings to specifying the claim name to be normalized to OIDC Connector. My PoC code is below:
https://github.com/summerwind/dex/blob/5547dc874b34d620d83062aba1e45f2ffa9a7015/connector/connector_oidc.go#L155-L172

What are your thoughts on?

**Comment by ericchiang** [comment]
Sorry for the delayed response. I've been out of the office.

I'm actually fine with this for non-required claims.

```
{
    "type": "oidc",
    "id": "azure",
    "issuerURL": "https://accounts.example.com",
    "clientID": "$CLIENT_ID",
    "clientSecret": "$CLIENT_SECRET",
    "trustedEmailProvider": true,
    "claimsMap": {
        "upn": "email"
    }
}
```

This get's trickier with the require claims. For example, that document[0] implies that Azure doesn't return a `sub` which OpenID Connect requires and used by dex.

Mapping kinda odd claims is fine but it looks like __Azure doesn't actually implement OpenID Connect__, only a subset. Maybe we should be thinking about an Azure connector?

[0] https://azure.microsoft.com/en-us/documentation/articles/guidance-multitenant-identity-claims/

**Comment by summerwind** [comment]
Thank you for the comment. 

> Mapping kinda odd claims is fine but it looks like Azure doesn't actually implement OpenID Connect, only a subset. Maybe we should be thinking about an Azure connector?

I totally agree with you. I'm going to implement an Azure connector in next week. I'll send a PR when it's done.

FYI: According to this docs, Azure AD seems to return a `sub` claim.
https://azure.microsoft.com/en-us/documentation/articles/active-directory-token-and-claims/#_subject

**Comment by summerwind** [comment]
Fixed in #527.
